Miscellaneous clean-ups and adjustments for PHP 8

Daimona opened the pull request:
See individual commits for more details.

Daimona committed:
Drop more old PHP references

Daimona committed:
Fix crash in PreferNamespaceUsePlugin and add test

In the polyfill, the return type node is (now) returnTypeList, so update
it accordingly to avoid the crash. Also don't bother looking at the
return type when fixing parameters (possibly a copypasta?). Fix a
copypasta in the plugin description and add a simple test.

Daimona committed:
Update types in ConstantInterface based on ConstantTrait

The trait had a more complete type list that was really meant for the
interface, so move it together with an explanatory comment, to avoid
signature mismatches.

Daimona committed:
Make PreferNamespaceUsePlugin handle union types

Daimona committed:
Autofix new issues from PreferNamespaceUsePlugin

Daimona committed:
Re-enable PHPDocToRealTypesPlugin and make pass

Mostly done automatically, but includes manual fixes to inaccurate doc
comments, or doc comments parsed incorrectly (as in
https://github.com/phan/phan/issues/5142).

Daimona committed:
Enable UnusedSuppressionPlugin and make pass

Most suppressions dropped automatically with an ad-hoc script, then
refined manually. Then autofixed the new issues that came up, like
redundant doc comments.

Daimona committed:
Fix test expectations for latest changes in v6

Daimona committed:
Fix suppression needed only in PHP < 8.3

rlerdorf reviewed (changes requested):
---

rlerdorf reviewed (dismissed):
---

rlerdorf reviewed (commented):
I see some self-analysis issues:
```
src/Phan/AST/TolerantASTConverter/TolerantASTConverter.php:3510 UnusedSuppression Element \Phan\AST\TolerantASTConverter\TolerantASTConverter::getDevelopmentBuildDate suppresses issue PhanImpossibleCondition but does not use it
src/Phan/AST/TolerantASTConverter/TolerantASTConverter.php:3512 PhanRedundantCondition Redundant attempt to cast \strpos(\PHP_VERSION, '-dev') of type 6 to truthy
src/Phan/polyfills/phan_ast_hash.php:16 PhanRedefineFunctionInternal Function phan_ast_hash defined at src/Phan/polyfills/phan_ast_hash.php:16 was previously defined internally
src/Phan/polyfills/phan_unique_types.php:13 PhanRedefineFunctionInternal Function phan_unique_types defined at src/Phan/polyfills/phan_unique_types.php:13 was previously defined internally
```

Daimona commented:
> I see some self-analysis issues:

Hmmm odd, I don't get any of those locally, and CI also passes (except appveyor, but I don't trust that). Could it be specific to your local environment? What PHP version are those from?

Daimona committed:
Remove suppressions for issue 5143, no longer needed

Daimona committed:
Update more references to PHP < 8.1

Drop dead code and descriptions that are now redundant. The version
numbers are no longer relevant for established features.

Daimona committed:
Drop test expectation normalization for PHP < 8

Daimona committed:
Throw an error when no files in include_analysis_file_list exist

Daimona committed:
Modernize phan_client

Stop trying to support PHP 5.6 (lol) and start using more modern syntax.

Daimona committed:
Restore test normalization for PHP <= 8.3

rlerdorf commented:
> > I see some self-analysis issues:
> 
> Hmmm odd, I don't get any of those locally, and CI also passes (except appveyor, but I don't trust that). Could it be specific to your local environment? What PHP version are those from?

The appVeyor failure is simple. It comes from `tests/multi_files/expected/551.php.expected` where you have:
```
./tests/multi_files/src/551_a.php:5 PhanUndeclaredConstant Reference to undeclared constant \CONSTANT_NOT_IN_SCOPE. This would throw an Error.
```
As in, you have the full path instead of the normal `%s` there, so on Windows it fails because of the forward slash vs. backslash issue

And I am not sure why you aren't seeing those self-analysis issues. Your PR  dropped some suppressions for actual issues. Are you scanning with v6? For example, in `src/Phan/AST/TolerantASTConverter/TolerantASTConverter.php` for `getDevelopmentBuildDate()` you changed the suppressions:
```
- @suppress PhanImpossibleTypeComparison, PhanRedundantCondition, PhanImpossibleCondition, PhanSuspiciousValueComparison Phan evaluates the strpos to a constant, so this is either impossible or redundant
+ @suppress PhanImpossibleCondition, PhanSuspiciousValueComparison Phan evaluates the strpos to a constant, so this is either impossible or redundant
```
which causes these with v6:
```
src/Phan/AST/TolerantASTConverter/TolerantASTConverter.php:3510 UnusedSuppression Element \Phan\AST\TolerantASTConverter\TolerantASTConverter::getDevelopmentBuildDate suppresses issue PhanImpossibleCondition but does not use it
src/Phan/AST/TolerantASTConverter/TolerantASTConverter.php:3512 PhanRedundantCondition Redundant attempt to cast \strpos(\PHP_VERSION, '-dev') of type 6 to truthy
```
And in `src/Phan/polyfills/phan_ast_hash.php` and `src/Phan/polyfills/phan_unique_types.php` your PR drops `@suppress PhanRedefineFunctionInternal` from the docblocks. I understand this doesn't trigger for you because you likely don't have the phan_helpers extension loaded, but it is there for people who do.

Daimona committed:
Fix remaining test failures

Daimona commented:
> The appVeyor failure is simple. [...] you have the full path instead of the normal `%s` there, so on Windows it fails because of the forward slash vs. backslash issue

Whoops I didn't realize I'd done that. I should trust appveyor more :D Fixed now.

> And I am not sure why you aren't seeing those self-analysis issues. Your PR dropped some suppressions for actual issues. Are you scanning with v6?

Yep. I can't reproduce this with either `__FakeSelfTest` (essentially `./phan`) or `__FakeSelfFallbackTest`. These also pass in CI.

> For example, in `src/Phan/AST/TolerantASTConverter/TolerantASTConverter.php` for `getDevelopmentBuildDate()` [...]

I should've read the doc comment of that method. It does a comparison on `PHP_VERSION`, but phan resolves the constant to a literal, so the comparison is either redundant or impossible depending on the PHP version. I suppose in your case it's redundant because you're using PHP8.5-dev while I'm using 8.4.x. Fixed by ignoring the unused suppression.

> And in `src/Phan/polyfills/phan_ast_hash.php` and `src/Phan/polyfills/phan_unique_types.php` your PR drops `@suppress PhanRedefineFunctionInternal` from the docblocks. I understand this doesn't trigger for you because you likely don't have the phan_helpers extension loaded, but it is there for people who do.

Ah I see. Using the dockerized test runner should make it more consistent, but at any rate, in the current version I'm ignoring unused suppressions for those two functions.

Daimona committed:
Drop config flag allow_method_param_type_widening

Parameter type contravariance is now an established PHP feature, so it's
not clear why someone would like to "disable" it. For the lack of real
types, one can use PHPDocToRealTypesPlugin or other linters.

Daimona committed:
Drop PHP53CompatibilityPlugin

I don't need to explain why, right?

rlerdorf commented:
Interesting, now appVeyor is failing on some `UnusedSuppression` warnings.  Why is that plugin even on in appVeyor?

Daimona committed:
Ignore errors from Phan phar in appveyor

This is only meant to test that the phar is working. Any issues it finds
(such as unused suppressions) should not fail the build.

Daimona commented:
> Interesting, now appVeyor is failing on some `UnusedSuppression` warnings. Why is that plugin even on in appVeyor?

I don't think that invocation should fail the build. It's just there to test that the phar is working. I changed it to always exit successfully regardless of what issues it finds.
